Adding a naive implementation of various primitive tensor operations

tannergooding opened the pull request.

tannergooding committed:
Adding a naive implementation of various primitive tensor operations

stephentoub commented on src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:
```diff
@@ -0,0 +1,467 @@
+            for (int i = 0; i < x.Length; i++)
+            {
+                destination[i] = x[i] + y[i];
+            }
```
I assume the plan is to get the functionality merged and to then follow-up with vectorization in time for 8.0 as well?

Let me know if you want to divvy that up at all or if you're going to tackle it all yourself.  Presumably we'll end up wanting a `Vector<T>` implementation for downlevel and a `Vector128/256/512<T>` for .NET 8+.

tannergooding replied:
Yes. The general plan is to get this first bit (covering element-wise operations) in.

Work on vectorizing them can then start while the vector-based tensor operations (Signmoid, Product, Sum, etc) get their naive implementations in.

@michaelgsharp and I will be splitting the work, but you're more than welcome to join in and help as well!

tannergooding replied:
Getting the naive implementation in gives us both a scalar fallback and a baseline against which correctness of the SIMD implementation can be validated.

I should have tests up here in just a little bit.

stephentoub replied:
> @michaelgsharp and I will be splitting the work, but you're more than welcome to join in and help as well!

Happy to help

stephentoub commented on src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.cs:
```diff
@@ -0,0 +1,467 @@
+            }
+        }
+    }
+}
```
This looks like it's only a subset of the approved surface. Is the plan to add the rest as part of this PR or as a follow-up?

tannergooding replied:
Answered in the comment thread just above.

tannergooding committed:
Adding tests covering the new tensor primitives APIs

tannergooding committed:
Adding tensor primitives APIs to the ref assembly

tannergooding committed:
Allow .NET Framework to build/run

ViktorHofer committed:
Sync TFMs between ref and src, csproj simplication and clean-up

ViktorHofer commented:
If the intent is to publicly ship this library going forward, you want to remove the properties in `C:\git\runtime4\src\libraries\System.Numerics.Tensors\Directory.Build.props` (and delete the file).

You will also want to update the README.md file which mentions that the library doesn't ship publicly.

michaelgsharp reviewed (approved):
Looks good to me. Probably still want another opinion though.

ViktorHofer commented on src/libraries/System.Numerics.Tensors/ref/System.Numerics.Tensors.csproj:
```diff
@@ -1,6 +1,7 @@
+
   <PropertyGroup>
-    <TargetFrameworks>$(NetCoreAppCurrent);$(NetCoreAppPrevious);$(NetCoreAppMinimum);netstandard2.0;$(NetFrameworkMinimum)</TargetFrameworks>
+    <TargetFrameworks>$(NetCoreAppCurrent);netstandard2.0;$(NetFrameworkMinimum)</TargetFrameworks>
```
[Now that `NetCoreAppCurrent` means net9.0](https://github.com/dotnet/runtime/commit/85fa924eab86aedd55dc751c8ef56cf2cc66d31), we should add a net8.0 minimum TFM as well.

```suggestion
    <TargetFrameworks>$(NetCoreAppCurrent);$(NetCoreAppPrevious);$(NetCoreAppMinimum);netstandard2.0;$(NetFrameworkMinimum)</TargetFrameworks>
```

ViktorHofer commented on src/libraries/System.Numerics.Tensors/src/System.Numerics.Tensors.csproj:
```diff
@@ -1,30 +1,28 @@
+  
   <PropertyGroup>
-    <TargetFrameworks>$(NetCoreAppCurrent);$(NetCoreAppPrevious);$(NetCoreAppMinimum);netstandard2.0;$(NetFrameworkMinimum)</TargetFrameworks>
+    <TargetFrameworks>$(NetCoreAppCurrent);netstandard2.0;$(NetFrameworkMinimum)</TargetFrameworks>
```
```suggestion
<TargetFrameworks>$(NetCoreAppCurrent);$(NetCoreAppPrevious);$(NetCoreAppMinimum);netstandard2.0;$(NetFrameworkMinimum)</TargetFrameworks>
```

stephentoub commented on src/libraries/System.Numerics.Tensors/tests/TensorPrimitivesTests.cs:
```diff
@@ -0,0 +1,633 @@
+{
+    public static class TensorPrimitivesTests
+    {
+        private const int TensorSize = 512;
```
Do we have any tests for boundary conditions, e.g. empty? Or error conditions?  Or is the plan for those to come separately?

tannergooding replied:
I added tests for the exception cases already. I can add ones for empty as well.

For floating-point correctness cases (i.e. `infinity`, `nan`, `-0`, etc), I think it'd be good to add handling for after we move off the "naive" implementation.

stephentoub commented on src/libraries/System.Numerics.Tensors/tests/TensorPrimitivesTests.cs:
```diff
@@ -0,0 +1,633 @@
+        [Fact]
+        public static void AddTwoTensors()
+        {
+            var x = CreateAndFillTensor(TensorSize);
```
Nit: var => float[]

tannergooding replied:
Fixed all usages of `var`

michaelgsharp committed:
Apply suggestions from code review

Co-authored-by: Viktor Hofer <viktor.hofer@microsoft.com>

ViktorHofer commented:
Please respond to my above question before merging: https://github.com/dotnet/runtime/pull/91228#issuecomment-1699705294

tannergooding committed:
Don't use var

tannergooding committed:
Fix the S.N.Tensors readme and remove the file marking it as non-shipping

tannergooding commented:
> Please respond to my above question before merging: https://github.com/dotnet/runtime/pull/91228#issuecomment-1699705294

Removed the file that marked it as non-shipping and adjusted the readme.

ViktorHofer commented on src/libraries/System.Numerics.Tensors/README.md:
```diff
@@ -1,2 +1,3 @@
-This library has not been shipped publicly and is not accepting contributions at this time.
\ No newline at end of file
+
+Provides APIs for performing primitive operations over tensors represented by spans of memory.
```
Would be good to update this README to the format that other libraries use, i.e. https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Configuration.Abstractions/README.md. This can happen in a follow-up but please don't forget about it.

Also, as this package is now shipping officially, we should add a PACKAGE readme as well. I will add an entry in https://github.com/dotnet/runtime/pull/91210 to track that work.
